File: collect_app/src/test/java/org/odk/collect/android/support/FakeScheduler.java
package org.odk.collect.android.support;

import org.jetbrains.annotations.NotNull;
import org.odk.collect.async.Cancellable;
import org.odk.collect.async.Scheduler;
import org.odk.collect.async.TaskSpec;

import java.util.function.Consumer;
import java.util.function.Supplier;

public class FakeScheduler implements Scheduler {

    private Runnable foregroundTask;
    private Runnable backgroundTask;
    private Boolean cancelled = false;
    private Boolean isRepeatRunning = false;

    @Override
    public <T> void immediate(Supplier<T> foreground, Consumer<T> background) {
        backgroundTask = () -> background.accept(foreground.get());
    }

    @Override
    public void immediate(@NotNull Runnable foreground) {
        foregroundTask = foreground;
    }

    @Override
    public void networkDeferred(@NotNull String tag, @NotNull TaskSpec spec) {

    }

    @Override
    public void networkDeferred(@NotNull String tag, @NotNull TaskSpec taskSpec, long repeatPeriod) {

    }

    @Override
    public Cancellable repeat(Runnable foreground, long repeatPeriod) {
        this.foregroundTask = foreground;
        isRepeatRunning = true;
        return () -> {
            isRepeatRunning = false;
            cancelled = true;
            return true;
        };
    }

    public void runForeground() {
        if (foregroundTask != null) {
            foregroundTask.run();
        }

    }

    public void runBackground() {
        if (backgroundTask != null) {
            backgroundTask.run();
        }
    }

    public Boolean hasBeenCancelled() {
        return cancelled;
    }

    public Boolean checkRepeatRunning() {
        return isRepeatRunning;
    }

    @Override
    public boolean isRunning(@NotNull String tag) {
        return false;
    }

    @Override
    public void cancelDeferred(@NotNull String tag) {

    }
}

File: testshared/src/main/java/org/odk/collect/testshared/FakeScheduler.kt
package org.odk.collect.testshared

import org.odk.collect.async.Cancellable
import org.odk.collect.async.Scheduler
import org.odk.collect.async.TaskSpec
import java.util.LinkedList
import java.util.function.Consumer
import java.util.function.Supplier

class FakeScheduler : Scheduler {

    private var foregroundTasks = LinkedList<Runnable>()
    private var backgroundTask: Runnable? = null
    private var repeatTasks = ArrayList<RepeatTask>()

    override fun <T> immediate(foreground: Supplier<T>, background: Consumer<T>) {
        backgroundTask = Runnable { background.accept(foreground.get()) }
    }

    override fun immediate(foreground: Runnable) {
        foregroundTasks.push(foreground)
    }

    override fun networkDeferred(tag: String, spec: TaskSpec, inputData: Map<String, String>) {}

    override fun networkDeferred(
        tag: String,
        taskSpec: TaskSpec,
        repeatPeriod: Long,
        inputData: Map<String, String>
    ) {
    }

    override fun repeat(foreground: Runnable, repeatPeriod: Long): Cancellable {
        foregroundTasks.add(foreground)

        val task = RepeatTask(repeatPeriod, foreground, 0)
        repeatTasks.add(task)

        return object : Cancellable {
            override fun cancel(): Boolean {
                repeatTasks.remove(task)
                return true
            }
        }
    }

    override fun cancelAllDeferred() {}

    fun runForeground() {
        while (foregroundTasks.isNotEmpty()) {
            foregroundTasks.remove().run()
        }

        repeatTasks.forEach { it.runnable.run() }
    }

    fun runForeground(currentTime: Long) {
        while (foregroundTasks.isNotEmpty()) {
            foregroundTasks.remove().run()
        }

        repeatTasks.forEach {
            if ((currentTime - it.lastRun) >= it.interval) {
                it.runnable.run()
                it.lastRun = currentTime
            }
        }
    }

    fun runBackground() {
        if (backgroundTask != null) {
            backgroundTask!!.run()
            backgroundTask = null
        }
    }

    fun isRepeatRunning(): Boolean {
        return repeatTasks.isNotEmpty()
    }

    override fun isDeferredRunning(tag: String): Boolean {
        return false
    }

    override fun cancelDeferred(tag: String) {}
}

private data class RepeatTask(val interval: Long, val runnable: Runnable, var lastRun: Long)
